File: apps/socket_io_events.py
from apps import _dict, socketio

def open_uri(uri):
    socketio.emit("open_new_tab", {"url": uri})

def show_alert(content):
    socketio.emit("dialog", {"type": "alert", "content": content})

def msgprint(msg="", title=None):
    socketio.emit("dialog", {"type": "msgprint", "msg": msg, "title": title})

def progress( count=0, total=100, title=None, description=""):
    socketio.emit("dialog", {
        "type": "progress", 
        "description": description, 
        "title": title,
        "count": count,
        "total":total
    })

def remove_tab_with_url(url):
    socketio.emit("remove_tab_with_url", {"url": url})

File: apps/static/assets/js/socketio_events.js
$(function(){
	const socket = io();

	socket.on('open_new_tab', (data) => {
		window.open(data.url);
	});

	socket.on("dialog", (data) =>{
		if (data.type == "progress"){
			modal_progress(data);
		}
		else if (data.type == "msgprint"){
			modal_msgprint(data);
		}
		else if (data.type == "alert"){
			alertToast(data.content);
		}
		
	});

});
